Sources/Datadog/InternalMonitoring/Kronos/KronosMonitor.swift

```diff
@@ -0,0 +1,294 @@
+    private let dispatchGroup = DispatchGroup()
+
+    /// The address of the pool being synchronised by `KronosClock`.
+    private var pool: String? = nil
+    /// The final server offset retrieved from `KronosClock` (can be `nil` if anything went wrong).
+    private var serverOffset: TimeInterval? = nil
+    /// Stats for each IP resolved from the `pool`.
+    private var statsByIP: [KronosInternetAddress: IPStats] = [:]
+    /// Connection monitor checking additional reachability for each IP resolved from the `pool`. Only available from iOS 14.2.
+    private let connectionMonitor: IPConnectionMonitorType?
+
+    /// Internal Monitor for sending the `result` to Datadog.
+    private var exporter: InternalMonitor? = nil
+
+    convenience init() {
+        let queue = DispatchQueue(label: "com.datadoghq.kronos-monitor", qos: .utility)
+        if #available(iOS 14.2, *) {
+            self.init(
+                queue: queue,
+                connectionMonitor: IPConnectionMonitor(queue: queue)
+            )
+        } else {
+            self.init(
+                queue: queue,
+                connectionMonitor: nil
+            )
+        }
+    }
+
+    init(queue: DispatchQueue, connectionMonitor: IPConnectionMonitorType?) {
+        self.queue = queue
+        self.connectionMonitor = connectionMonitor
+
+        self.dispatchGroup.enter() // await Kronos completion
+        self.dispatchGroup.enter() // await exporter registration
+
+        self.dispatchGroup.notify(queue: self.queue) { [weak self] in
+            self?.sendTelemetryToDatadog()
+        }
+    }
+
+    // MARK: - Clock sync
+
+    func notifySyncStart(from pool: String) {
+        queue.async {
+            self.pool = pool
+        }
+    }
+
+    func notifySyncEnd(serverOffset: TimeInterval?) {
+        queue.async {
+            self.serverOffset = serverOffset
+            self.dispatchGroup.leave() // notify Kronos completion
+        }
+    }
+
+    // MARK: - Exporting
+
+    func export(to exporter: InternalMonitor) {
+        queue.async {
+            self.exporter = exporter
+            self.dispatchGroup.leave() // notify exporter registration
+        }
+    }
+
+    // MARK: - DNS resolution
+
+    func notifyResolveDNS(to addresses: [KronosInternetAddress]) {
+        queue.async {
+            addresses.forEach { ip in
+                self.statsByIP[ip] = IPStats()
+
+                if let connectionMonitor = self.connectionMonitor {
+                    self.dispatchGroup.enter() // await connection check result
+                    connectionMonitor.checkConnection(to: ip) { [weak self] result in // completion is called on `queue`
+                        self?.statsByIP[ip]?.checkResult = result
+                        self?.dispatchGroup.leave() // notify connection check result
+                    }
+                }
+            }
```

Comment:
```swift
guard let connMonitor = self.connectionMonitor else { return }
addresses.forEach { _ in self.dispatchGroup.enter() }
addresses.forEach { ip in
  self.statsByIP[ip] = IPStats()
  connMonitor.checkConnection(to: ip) { [weak self] result in
    self?.statsByIP[ip]?.checkResult = result
    self?.dispatchGroup.leave()
  }
}
```
that way if `checkConnection` calls its `completion` too early somehow, we can avoid `dispatchGroup` being left too early as well

Reply:
I see the point 👍 and this would make sense if completion is ran on an arbitrary thread:

```swift
queue.async {
    addresses.forEach { _ in
        if let _ = self.connectionMonitor { // ⚠️ nil on iOS < 14.2
            self.dispatchGroup.enter() // await connection check result
        }
    }

    addresses.forEach { ip in
        self.statsByIP[ip] = IPStats()

        if let connectionMonitor = self.connectionMonitor { // ⚠️ nil on iOS < 14.2
            connectionMonitor.checkConnection(to: ip) { [weak self] result in // completion is called on `queue`
                self?.statsByIP[ip]?.checkResult = result
                self?.dispatchGroup.leave() // notify connection check result
            }
        }
    }
}

```

However in current model it is impossible to receive `completion` simultaneously. Both `checkConnection(to:completion:)` and its `completion:` are called on the same queue, meaning that both "start all checks" and "receive single completion" are part of the same critical section and can't happen in parallel as stated:

```swift
connectionMonitor.checkConnection(to: ip) { [weak self] result in // completion is called on `queue`
```

In fact, if that wouldn't been assured, we'd also have problem with thread safe access to `self.statsByIP` from `completion`, whereas it works fine and is covered in tests.